### PR TITLE
[Early review only - DO NOT MERGE] docker image reference types - WIP

### PR DESCRIPTION
**Pull request by glyn**
Fixes https://github.com/projectriff/riff/issues/864

### CODE REVIEW

**Comment by glyn** [comment]
This PR is just to surface my current work until I return from vacation. The code is far from complete and many tests fail. The changes are limited in scope and need to be expanded out so that images are converted to `Reference` types as soon as they are encountered by the `core` package.

The big change is that normalisation ensures that some of the edge cases previously coded for will no longer occur. This is the reason that many of the tests fail.

Any comments gratefully received.

File: pkg/core/image.go [hunk 1]
```diff
@@ -239,8 +241,12 @@ func relocateImageManifest(imageManifestPath string, mapper *imageMapper, output
...
 
 func applyMapper(name imageName, mapper *imageMapper) imageName {
-	quotedName := imageName(mapper.mapImages([]byte(fmt.Sprintf("%q", name))))
-	return quotedName[1 : len(quotedName)-1]
+	mapped := string(mapper.mapImages([]byte(fmt.Sprintf("%q", name.String()))))
+	ref, err := reference.Parse(mapped[1 : len(mapped)-1])
+	if err != nil {
+		panic(err) // TODO: handle this gracefully
+	}
+	return imageName(ref)
```
Comment (ericbottard) [inline pkg/core/image.go]: I had already forgotten that we already had an `imageName` (opaque) type...
